File: DAY01/distances.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include "mapImplementation.h"

int main() {
    std::ifstream infile("/Users/julesdupont/school/aoc24/DAY01/data.txt");
    if (!infile) {
        std::cerr << "Unable to open file";
        return 1;
    }

    std::vector<int> leftVector, rightVector;
    int left, right;

    while (infile >> left >> right) {
        leftVector.push_back(left);
        rightVector.push_back(right);
    }

    infile.close();

    std::sort(leftVector.begin(), leftVector.end());
    std::sort(rightVector.begin(), rightVector.end());

    int distancesSum = 0;

    for (int i = 0; i < leftVector.size(); i++) {
        int distance = std::abs(leftVector.at(i) - rightVector.at(i));
        distancesSum += distance;
    }

    printf("%d\n", distancesSum);

    // PART 2
    mapImplementation map;
    for (int i = 0; i < leftVector.size(); i++) {
        if (!map.isKey(leftVector.at(i))) {
            map.insert(leftVector.at(i), rightVector.at(i));
        }
    }


    return 0;
}

File: DAY01/distances.py
with open('data.txt') as f:
    lines = f.readlines()

leftList, rightList = [], []

for line in lines:
    leftString, rightString = '', ''
    i = 0
    while line[i] != ' ':
        leftString += line[i]
        i += 1
    i += 3
    while i < len(line):
        if line[i] == '\n':
            break
        rightString += line[i]
        i += 1
    leftList.append(int(leftString))
    rightList.append(int(rightString))

sortedLeftList = sorted(leftList)
sortedRightList = sorted(rightList)

distances = []

for i in range(len(lines)) :
    left = sortedLeftList[i]
    right = sortedRightList[i]
    distances.append(abs(left - right))

# ANSWER FOR PART 1
print(sum(distances))



occurences = {}
for i in range(len(leftList)) :
    if leftList[i] not in occurences:
        s = 0
        for j in range(len(rightList)) :
            s += 1 if (rightList[j] == leftList[i]) else 0
        occurences[leftList[i]] = s

similarity = 0
for i in range(len(leftList)) :
    similarity += leftList[i] * occurences[leftList[i]]

# ANSWER FOR PART 2
print(similarity)